File: ruby/GrabzIt/lib/grabzit/videooptions.rb
module GrabzIt
	require File.join(File.dirname(__FILE__), 'baseoptions')
	
	# Represents all of the options available when creating an video
	# @version 3.0
	# @author GrabzIt
	class VideoOptions < BaseOptions
		def initialize()
			super()
			@browserWidth = nil
			@browserHeight = nil
			@waitForElement = nil
			@requestAs = 0
			@customWaterMarkId = nil
			@noAds = false
			@noCookieNotifications = false
			@address = nil
			@clickElement = nil
			@start = 0
			@duration = 10
			@framesPerSecond = 0
		end
		
		# @return [Integer] the width of the browser in pixels
		def browserWidth
			@browserWidth
		end
		
		# Set the width of the browser in pixels
		#
		# @param value [Integer] the browser width
		# @return [void]		
		def browserWidth=(value)
			@browserWidth = value
		end	
		
		# @return [Integer] the height of the browser in pixels
		def browserHeight
			@browserHeight
		end
		
		# Set the height of the browser in pixels. Use -1 to screenshot the whole web page
		#
		# @param value [Integer] the browser height
		# @return [void]		
		def browserHeight=(value)
			@browserHeight = value
		end
		
		# @return [String] get the CSS selector of the HTML element in the web page that must be visible before the capture is performed
		def waitForElement
			@waitForElement
		end
		
		# Set the CSS selector of the HTML element in the web page that must be visible before the capture is performed
		#
		# @param value [String] the element to wait for
		# @return [void]		
		def waitForElement=(value)
			@waitForElement = value
		end
		
		# @return [String] get the CSS selector of the HTML element in the web page that must clicked before the capture is performed
		def clickElement
			@clickElement
		end
		
		# Set the CSS selector of the HTML element in the web page that must clicked before the capture is performed
		#
		# @param value [String] the element to click
		# @return [void]		
		def clickElement=(value)
			@clickElement = value
		end		
		
		# @return [Integer] get which user agent type should be used
		def requestAs
			@requestAs
		end
		
		# Set which user agent type should be used: Standard Browser = 0, Mobile Browser = 1, Search Engine = 2 and Fallback Browser = 3
		#
		# @param value [Integer] the browser type
		# @return [void]		
		def requestAs=(value)
			@requestAs = value
		end		
		
		# @return [String] the custom watermark id.
		def customWaterMarkId
			@customWaterMarkId
		end
		
		# Set a custom watermark to add to the screenshot.
		#
		# @param value [String] custom watermark id
		# @return [void]		
		def customWaterMarkId=(value)
			@customWaterMarkId = value
		end
		
		# @return [Boolean] get if adverts should be automatically hidden
		def noAds
			@noAds
		end
		
		# Set to true if adverts should be automatically hidden
		#
		# @param value [Boolean] hide adverts
		# @return [void]		
		def noAds=(value)
			@noAds = value
		end		

		# @return [Boolean] get if cookie notifications should be automatically hidden
		def noCookieNotifications
			@noCookieNotifications
		end
		
		# Set to true if cookie notifications should be automatically hidden
		#
		# @param value [Boolean] hide cookie notifications
		# @return [void]		
		def noCookieNotifications=(value)
			@noCookieNotifications = value
		end

		# @return [String] get the URL to execute the HTML code in
		def address
			@address
		end
		
		# Set the URL to execute the HTML code in
		#
		# @param value [String] the address
		# @return [void]		
		def address=(value)
			@address = value
		end
		
		# @return [Integer] the starting time of the web page that should be converted into a video
		def start
			@start
		end
		
		# Set the starting time of the web page that should be converted into a video
		#
		# @param value [Integer] the second to start at
		# @return [void]		
		def start=(value)
			@start = value
		end			

		# @return [Integer] the length in seconds of the web page that should be converted into a video
		def duration
			@duration
		end
		
		# Set the length in seconds of the web page that should be converted into a video
		#
		# @param value [Integer] the number of seconds
		# @return [void]		
		def duration=(value)
			@duration = value
		end		
		
		# @return [Float] the number of frames per second that should be used to create the video. From a minimum of 0.2 to a maximum of 10
		def framesPerSecond
			@framesPerSecond
		end
		
		# Set the number of frames per second that should be used to create the video. From a minimum of 0.2 to a maximum of 10
		#
		# @param value [Float] the number of frames per second
		# @return [void]		
		def framesPerSecond=(value)
			@framesPerSecond = value
		end			
		
		# Define a HTTP Post parameter and optionally value, this method can be called multiple times to add multiple parameters. Using this method will force 
		# GrabzIt to perform a HTTP post.
		#
		# @param name [String] the name of the HTTP Post parameter
		# @param value [String] the value of the HTTP Post parameter
		def add_post_parameter(name, value)
			@post = appendParameter(@post, name, value)
		end		
		
		# @!visibility private
		def _getSignatureString(applicationSecret, callBackURL, url = nil)
			items = [applicationSecret]
			
			if(url != nil)
				items.push(GrabzIt::Utility.nil_check(url))
			end
			
			items.push(GrabzIt::Utility.nil_check(callBackURL),
			GrabzIt::Utility.nil_int_check(@browserHeight),
			GrabzIt::Utility.nil_int_check(@browserWidth),
			GrabzIt::Utility.nil_check(@customId),
			GrabzIt::Utility.nil_check(@customWaterMarkId),
			GrabzIt::Utility.nil_int_check(@start),
			GrabzIt::Utility.nil_int_check(@requestAs),
			GrabzIt::Utility.nil_check(@country),
			GrabzIt::Utility.nil_check(@exportURL),	
			GrabzIt::Utility.nil_check(@waitForElement),
			GrabzIt::Utility.nil_check(@encryptionKey),
			GrabzIt::Utility.b_to_str(@noAds),
			GrabzIt::Utility.nil_check(@post),
			GrabzIt::Utility.nil_check(@proxy),
			GrabzIt::Utility.nil_check(@address),
			GrabzIt::Utility.b_to_str(@noCookieNotifications),
			GrabzIt::Utility.nil_check(@clickElement),
			GrabzIt::Utility.nil_float_check(@framesPerSecond),
			GrabzIt::Utility.nil_int_check(@duration))
			
			return items.join("|")
		end
	
		# @!visibility private
		def _getParameters(applicationKey, sig, callBackURL, dataName, dataValue)
			params = createParameters(applicationKey, sig, callBackURL, dataName, dataValue)		
			params['bwidth'] = GrabzIt::Utility.nil_int_check(@browserWidth)
			params['customwatermarkid'] = GrabzIt::Utility.nil_check(@customWaterMarkId)
			params['bheight'] = GrabzIt::Utility.nil_int_check(@browserHeight)
			params['waitfor'] = GrabzIt::Utility.nil_check(@waitForElement)
			params['requestmobileversion'] = GrabzIt::Utility.nil_int_check(@requestAs)		
			params['noads'] = GrabzIt::Utility.b_to_str(@noAds)
			params['post'] = GrabzIt::Utility.nil_check(@post)
			params['address'] = GrabzIt::Utility.nil_check(@address)
			params['nonotify'] = GrabzIt::Utility.b_to_str(@noCookieNotifications)
			params['click'] = GrabzIt::Utility.nil_check(@clickElement)
			params['start'] = GrabzIt::Utility.nil_int_check(@start)
			params['fps'] = GrabzIt::Utility.nil_float_check(@framesPerSecond)
			params['duration'] = GrabzIt::Utility.nil_int_check(@duration)

			return params
		end
	end
end

File: aspnet/GrabzIt/Parameters/VideoOptions.cs
using System;
using System.Collections.Generic;
using System.Reflection.Emit;
using System.Runtime.InteropServices;
using System.Text;
using GrabzIt.COM;
using GrabzIt.Enums;

namespace GrabzIt.Parameters
{
    [ClassInterface(ClassInterfaceType.None)]
    public class VideoOptions : BaseOptions, IVideoOptions
    {
        public VideoOptions()
        {
            RequestAs = BrowserType.StandardBrowser;
            CustomWaterMarkId = string.Empty;
            BrowserWidth = 0;
            BrowserHeight = 0;
            Duration = 10;
            Address = string.Empty;
            ClickElement = string.Empty;
            JSCode = string.Empty;
        }

        /// <summary>
        /// The width of the browser in pixels
        /// </summary>
        public int BrowserWidth
        {
            get;
            set;
        }

        /// <summary>
        /// The height of the browser in pixels
        /// </summary>
        public int BrowserHeight
        {
            get;
            set;
        }

        /// <summary>
        /// The CSS selector(s) of the HTML element in the web page to click
        /// </summary>
        public string ClickElement
        {
            get;
            set;
        }

        /// <summary>
        /// The CSS selector of the HTML element in the web page that must be visible before the capture is performed
        /// </summary>
        public string WaitForElement
        {
            get;
            set;
        }

        /// <summary>
        /// Request screenshot in different forms: Standard Browser, Mobile Browser and Search Engine
        /// </summary>
        public BrowserType RequestAs
        {
            get;
            set;
        }

        /// <summary>
        /// Add a custom watermark to the image
        /// </summary>
        public string CustomWaterMarkId
        {
            get;
            set;
        }

        /// <summary>
        /// True if adverts should be automatically hidden.
        /// </summary>
        public bool NoAds
        {
            get;
            set;
        }

        /// <summary>
        /// True if cookie notification should be automatically hidden.
        /// </summary>
        public bool NoCookieNotifications
        {
            get;
            set;
        }

        /// <summary>
        /// The URL to execute the HTML code in.
        /// </summary>
        public string Address
        {
            get;
            set;
        }

        /// <summary>
        /// The starting time of the web page that should be converted into a video
        /// </summary>
        public int Start
        {
            get;
            set;
        }

        /// <summary>
        /// The length in seconds of the web page that should be converted into a video
        /// </summary>
        public int Duration
        {
            get;
            set;
        }

        /// <summary>
        /// The number of frames per second that should be used to create the video. From a minimum of 0.2 to a maximum of 10
        /// </summary>
        public float FramesPerSecond
        {
            get;
            set;
        }

        /// <summary>
        /// The height of the resulting video in pixels
        /// </summary>
        public int OutputHeight
        {
            get;
            set;
        }

        /// <summary>
        /// The width of the resulting video in pixels
        /// </summary>
        public int OutputWidth
        {
            get;
            set;
        }

        /// <summary>
        /// The JavaScript code that will be execute in the web page before the capture is performed
        /// </summary>
        public string JSCode
        {
            get;
            set;
        }

        /// <summary>
        /// Define a HTTP Post parameter and optionally value, this method can be called multiple times to add multiple parameters. Using this method will force 
        /// GrabzIt to perform a HTTP post.
        /// </summary>
        /// <param name="name">The name of the HTTP Post parameter</param>
        /// <param name="value">The value of the HTTP Post parameter</param>
        public void AddPostParameter(string name, string value)
        {
            post = AppendParameter(post, name, value);
        }

        internal override string GetSignatureString(string applicationSecret, string callBackURL, string url)
        {
            string urlParam = string.Empty;
            if (!string.IsNullOrEmpty(url))
            {
                urlParam = url + "|";
            }

            string callBackURLParam = string.Empty;
            if (!string.IsNullOrEmpty(callBackURL))
            {
                callBackURLParam = callBackURL;
            }

            return applicationSecret + "|" + urlParam + callBackURLParam +
            "|" + BrowserHeight
            + "|" + BrowserWidth + "|" + CustomId 
            + "|" + CustomWaterMarkId + "|" + Start + "|" + ((int)RequestAs).ToString() + "|" + ConvertCountryToString(Country) 
            + "|" + ExportURL + "|" + WaitForElement + "|" + EncryptionKey
             + "|" + Convert.ToInt32(NoAds) + "|" + post + "|" + Proxy + "|" + Address + "|" + Convert.ToInt32(NoCookieNotifications)
             + "|" + ClickElement + "|" + FramesPerSecond + "|" + Duration + "|" + OutputWidth + "|" + OutputHeight + "|" + JSCode;
        }

        protected override Dictionary<string, string> GetParameters(string applicationKey, string signature, string callBackURL, string dataName, string dataValue)
        {
            Dictionary<string, string> parameters = CreateParameters(applicationKey, signature, callBackURL, dataName, dataValue);
            parameters.Add("bwidth", BrowserWidth.ToString());
            parameters.Add("customwatermarkid", CustomWaterMarkId);
            parameters.Add("bheight", BrowserHeight.ToString());
            parameters.Add("start", Start.ToString());
            parameters.Add("waitfor", WaitForElement);
            parameters.Add("requestmobileversion", Convert.ToInt32(RequestAs).ToString());
            parameters.Add("noads", Convert.ToInt32(NoAds).ToString());
            parameters.Add("post", post);
            parameters.Add("address", Address);
            parameters.Add("nonotify", Convert.ToInt32(NoCookieNotifications).ToString());
            parameters.Add("click", ClickElement);
            parameters.Add("fps", FramesPerSecond.ToString());
            parameters.Add("duration", Duration.ToString());
            parameters.Add("width", OutputWidth.ToString());
            parameters.Add("height", OutputHeight.ToString());
            parameters.Add("jscode", JSCode);

            return parameters;
        }
    }
}
